File: src/generator.ts
import { SvgAST } from './transformer';

export function generator(svgAst: SvgAST): string {
  const createAttributes = (attr) => {
    return Object.keys(attr).map((key) => {
      return `${key}="${attr[key]}"`;
    }).join(' ');
  };

  const attributes = createAttributes(svgAst.attr);

  const elements = svgAst.body.map((node) => {
    return `<${node.tag} ${createAttributes(node.attr)}></${node.tag}>`;
  }).join('');

  return `<svg ${attributes}>${elements}</svg>`;
}

File: ruby/lib/sbn_compiler_rb/generator.rb
module SbnCompilerRb
  class Generator
    def generate(sbn_ast)
      attributes = build_attributes(sbn_ast.attr)
      elements = build_elements(sbn_ast.body)
      build_tag(tag: "svg", attributes: attributes, elements: elements)
    end

    private

    def build_attributes(sbn_ast_attr)
      sbn_ast_attr.map { |k, v| "#{k}=\"#{v}\"" }.join(' ')
    end

    def build_elements(sbn_ast_body)
      sbn_ast_body.map { |expressoin|
        tag = expressoin.tag
        attributes = build_attributes(expressoin.attr)
        build_tag(tag: tag, attributes: attributes)
      }.join
    end

    def build_tag(tag:, attributes:, elements: "")
      "<#{tag} #{attributes}>#{elements}</#{tag}>"
    end
  end
end
